Add notes on indexing to kNN search guide (#83188)
elastic/elasticsearch: docs/reference/search/search-your-data/knn-search.asciidoc

vector function

In most cases, you'll want to use approximate kNN. Approximate kNN offers lower

> at the cost of slower indexing and imperfect accuracy.

Reviewer: lower latency?

Author: Oops, thanks for catching that.

each shard to return the global top `k` nearest neighbors.

You can increase `num_candidates` for more accurate results at the cost of
slower search speeds. A search with a high value for `num_candidates`
considers more candidates from each shard. This takes more time, but the
search has a higher probability of finding the true `k` top nearest neighbors.

Similarly, you can decrease `num_candidates` for faster searches with
potentially less accurate results.

[discrete]
[[knn-indexing-considerations]]

> ==== Indexing considerations

Author: I'm very open to suggestions for a better name :)

Reviewer: What is the reason we put this section in `knnSearch guide`? Isn't a better place for it  the indexing section of `dense-vector` document?
Or we provide these kind of suggestions only in reference guides as this one?

Author: I was wondering this myself and would appreciate @jrodewig's opinion here. I was thinking of this reference as a combined "how to" for kNN search, where we offer guidance about what method to choose, other best practices, etc. I think of the API and field type docs more as specific references that don't give high-level guidance like this.

Indexing vectors for approximate kNN search can take substantial time because

> of how expensive it is to build the ANN index structures. You may need to

Reviewer: maybe, provide a definition in brackets what is ANN (approximate NN)?

Author: Good point, I'll use a different phrase since we don't use the acronym ANN in this reference.

> Indexing vectors for approximate kNN search can take substantial time because
> of how expensive it is to build the ANN index structures. You may need to
> increase the client request timeout for index and bulk requests.

Reviewer 2: Instead of `ANN`, we may want to mention HNSW graphs here. We talk about them in the next paragraph without much of an introduction.

I took a stab at a suggestion, but feel free to edit or ignore if wanted.

```suggestion
{es} shards are composed of segments, which are internal storage elements in the
index. For approximate kNN search, {es} stores the dense vector values of each
segment as an https://arxiv.org/abs/1603.09320[HNSW graph]. Indexing vectors for
approximate kNN search can take substantial time because of how expensive it is
to build these graphs. You may need to increase the client request timeout for
index and bulk requests.
```

Author: This reorganization works nicely, I'll adopt it.

> Elasticsearch shards are composed of segments, which are internal storage
> elements in the index. <<indices-forcemerge,Force merging>> the index to a

Reviewer 2: If you take my above suggestion, you don't need the first sentence of this paragraph.
```suggestion
<<indices-forcemerge,Force merging>> the index to a
```

single segment can improve kNN search latency. With only one segment, the

> search needs to check a single, all-inclusive HNSW graph. But when there are

Reviewer 2: Nit: unneeded word.
```suggestion
search needs to check a single, all-inclusive HNSW graph. When there are
```

multiple segments, kNN search must check several smaller HNSW graphs as it

> searches each segment after another. Note that you should only force merge an
> index if it is no longer being written to.

Reviewer 2: Minor rewording suggestion. Feel free to ignore.
```suggestion
searches each segment after another. Only force merge an index if it no longer
receives writes.
```
